src/CoreShop/Component/Order/Model/ProposalPriceRuleTrait.php

```diff
@@ -101,23 +114,71 @@ public function removePriceRule($priceRule)
     /**
      * {@inheritdoc}
      */
-    public function hasPriceRule($priceRule)
+    public function hasPriceRule(ProposalCartPriceRuleItemInterface $priceRule)
     {
         $items = $this->getPriceRuleItems();
 
         if ($items instanceof Fieldcollection) {
             foreach ($items as $item) {
-                if ($item instanceof ProposalCartPriceRuleItem) {
-                    if ($item->getCartPriceRule() instanceof CartPriceRuleInterface) {
-                        if ($item->getCartPriceRule()->getId() === $priceRule->getId()) {
-                            return true;
-                        }
-                    }
+                if (!$item instanceof ProposalCartPriceRuleItem) {
+                    continue;
+                }
+
+                if (!$item->getCartPriceRule() instanceof CartPriceRuleInterface) {
+                    continue;
+                }
+
+                if (!$priceRule->getCartPriceRule() instanceof CartPriceRuleInterface) {
+                    continue;
+                }
+
+                if ($item->getCartPriceRule()->getId() === $priceRule->getCartPriceRule()->getId()
+                    && $item->getVoucherCode() === $priceRule->getVoucherCode()) {
+                    return true;
                 }
             }
         }
 
         return false;
     }
 
+    /**
+     * {@inheritdoc}
+     */
+    public function hasCartPriceRule(
+        CartPriceRuleInterface $priceRule,
+        CartPriceRuleVoucherCodeInterface $voucherCode = null
+    ) {
+        return null !== $this->getPriceRuleByCartPriceRule($priceRule, $voucherCode);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getPriceRuleByCartPriceRule(
```

Comment:
this functions check if there is already a Fieldcollection Item for a given set of Cart Rule and Voucher